status-im/status-react
# Supported locales and translation ids

Fixes #2456

### User Stories

#### User Story 1
As a translator I want to ensure that I translate all English labels for the locale I am translated to be supported if that language is considered supported. If I don't, tests should fail, and I should be able to see what labels are missing from the tests report.

#### User Story 2

As a translator I want to be able to send translations for a locale, even if I don't send all translations for that locale to be supported, if that locale is not considered supported. Tests should not fail if I do so.

#### User story 3

As a reviewer I want to ensure that when a locale is considered supported in a PR, it has all translations to be really supported. Tests should fail otherwise, and I should be able to see that labels are missing in the tests report to tell them to the PR creator.

#### User story 4

As a reviewer I want to ensure that when translations are added, they are added to labels that exist in English. Tests should fail otherwise, and I should be able to see what labels don't exist in the tests report to tell them to the PR creator.

#### User story 5

As X I want to ensure that when I level up the checkpoint that specifies what labels are needed for a locale to be supported, all locales specified as supported actually pass that checkpoint. Tests should fail otherwise and I should see in the tests report what locales are not ready for those checkpoints, and the labels that are missing, so that I can ask for translations.

#### User story 6

As someone concerned with UX, I want to see what labels are missing in supported languages compared to English. Tests should not fail though, since tests should only fail if supported locale don't pass the current checkpoint.

### Steps to test

1. Run `lein test-cljs` and notice tests pass.
2. Add an unsupported language (for example `:ja`) to supported language list (var `supported-locales` in file `src/status_im/i18n.cljs`) and notice that tests fail. Undo the change.
3. Add a new label (for example `:foobar`) to English translations (file `src/status_im/translations/en.cljs`), and notice tests pass.
4. Define a new checkpoint using the label from step 3 (for example  `::checkpoint-0-9-13 (set/union checkpoint-0-9-12-labels #{:foobar}`, added in var `checkpoints-def` of file `src/status_im/i18n.cljs`), and notice that tests pass.
5. Make the checkpoint in step 4 the required one (by changing the var `checkpoint-to-consider-locale-supported` in file `src/status_im/i18n.cljs`), and notice that tests fail for languages where it previously passed, since they do not have the new locale that now is in the required checkpoint. Notice that  you can see the exact translations that are missing and in what languages.
6. Add the translation `:foobar` to language `:de` (a supported language), and notice that tests continue failing, but the report does not include the `:de` locale now.

### Status

Ready.

### Additional

I added specs for some i18n things:

* trans-id/s
* locale/s
* supported-locale/s
* checkpoint/s

**Translation ids**

English is used as the source of truth, that is, to know the translation ids that really exist.

So I added a test to check whether other locales have extra translations that actually do not exist. This way we can easily keep translation ids consistent with the original ones.

It seems that right now there are quite a lot of translation ids in other languages that are not in English, so the test is failing.

**Supported locales**

The biggest concern was to know what locales we support and to give feedback to translators and ourselves about how supported they really are.

So firstly I added a var to specify what locales we expect to be supported, `supported-locales`.

The support we expect for those locales isn’t full, that is, all translations, but it can be partial. So I added to specify incremental checkpoints to define what translations should be present for a language to support a checkpoint.

The fn `locales-with-checkpoint` tells us the checkpoint for every locale, or `nil` it it doesn’t support any.

Once we have checkpoints, we can specify the one that is required for a language to be really supported. This is done in the var `checkpoint-to-consider-locale-supported`.

And finally, with the expected supported locales and the checkpoint we expect for those languages, I added a test to check whether those locales really support that checkpoint.

We can also check whether there are locales that are supported and we do not have them as supported with the fn `supported-locales-that-are-not-considered-supported`.

This way we are not very strict with translators (they can even add a new locale with just one translation without tests failing), but we check what we really care about: the supported languages we define against the checkpoints we define. And if a locale that was supported becomes unsupported we notice from the test and can ask for missing translations or unmark it as supported.

What checkpoints and supported locales should we have right now?

Let me know if there is anything I am missing.
